### PR TITLE
Resolve module paths as relative to appRootDir

### PR DESCRIPTION
**Pull request by PradnyaBaviskar**
In context of - https://github.com/strongloop/loopback-boot/issues/73, made following changes...
 1. Added test case to verify that a call to <br/>
```app.boot(app, { appRootDir: __dirname, bootScripts: ["./custom-boot/script" ] })```
<br/> resolves boot script as ``` __dirname+"/custom-boot/script.js"```

@bajtos, please review.

### CODE REVIEW

File: lib/compiler.js [hunk 1]
```diff
@@ -323,8 +319,15 @@ function resolveSourceDir(rootDir, sourceDir) {
       [0];
     if (srcDir)
       return srcDir;
+  } else {
+    try {
+      var resolved = path.resolve(relativePath);
```
Comment (bajtos) [inline lib/compiler.js]: Please give `resolved` a better, more descriptive name. E.g. `absolutePath` or `fullPath`.

Add a comment explaining why we need to call `require.resolve`.

File: lib/compiler.js [hunk 2]
```diff
@@ -298,13 +294,13 @@ function findModelDefinitions(rootDir, sources) {
...
 }
 
-function resolveSourceDir(rootDir, sourceDir) {
-  var srcDir = path.resolve(rootDir, sourceDir);
+function resolveAppPath(rootDir, relativePath) {
+  var srcDir = path.resolve(rootDir, relativePath);
   if (fs.existsSync(srcDir))
     return srcDir;
 
   // Handle module-relative path, e.g. `loopback/common/models`
```
Comment (bajtos) [inline lib/compiler.js]: This comment feels misleading now, because the code below handles both relative paths and module-relative paths now. I am proposing to move the comment to the "if" block below, and add a comment to the "else" block too.

File: test/compiler.test.js [hunk 3]
```diff
@@ -405,6 +405,17 @@ describe('compiler', function() {
       expect(instructions.files.boot).to.eql([initJs]);
     });
 
+    it('should resolve js files in `bootScripts`', function() {
```
Comment (bajtos) [inline test/compiler.test.js]: The test name IMO does not describe the scenario well. The only relevant distinction of this test is that the bootScript file does not provide extension, right? My proposal: "resolves missing file extensions in `bootScritps`", but feel free to come up with your own name.

In the modules I start from scratch, I prefer to use the present tense in test names instead of starting each test name with `should` - [DRY](http://en.wikipedia.org/wiki/Don%27t_repeat_yourself). (Even though older loopback modules use the "should" convention.) Please keep the test names consistent with the convention used in each project.

File: lib/compiler.js [hunk 4]
```diff
@@ -313,18 +310,25 @@ function resolveSourceDir(rootDir, sourceDir) {
...
       .filter(function(candidate) {
         return fs.existsSync(candidate);
       })
       [0];
-    if (srcDir)
-      return srcDir;
+    if (fullPath)
+      return fullPath;
+  } else {
+    // Handle relative path, e.g. `{appRootDir}/common/models`
```
Comment (bajtos) [inline lib/compiler.js]: Nitpick: the relative path is usually specified as "./common/models"

File: lib/compiler.js [hunk 5]
```diff
@@ -313,18 +310,25 @@ function resolveSourceDir(rootDir, sourceDir) {
...
+    try {
+      fullPath = require.resolve(fullPath);
+      return fullPath;
+    } catch (err) {
+      debug ('Skipping %s - %s', fullPath, err);
+    }
   }
-
```
Comment (bajtos) [inline lib/compiler.js]: Nitpick: please keep the empty line as a visual delimiter of different code blocks.

File: lib/compiler.js [hunk 6]
```diff
@@ -505,7 +509,12 @@ function resolveRelativePaths(relativePaths, appRootDir) {
   relativePaths.forEach(function(relativePath, k) {
     var start = relativePath.substring(0, 2);
     if (start === './' || start === '..') {
-      relativePaths[k] = path.resolve(appRootDir, relativePath);
+      var resolvedPath = resolveAppPath(appRootDir, relativePath);
+      if (resolvedPath !== undefined) {
+        relativePaths[k] = resolvedPath;
+      } else {
+        debug ('skipping boot script %s - unknown file', relativePath);
+      }
```
Comment (bajtos) [inline lib/compiler.js]: Is this covered by a unit test too?

Also I think you can remove the check `if (start === './' || start === '..') {`now, because `resolveAppPath` correctly handles non-relative paths too, or not? Please add a unit-test for this too.

File: lib/compiler.js [hunk 7]
```diff
@@ -149,8 +149,8 @@ function findScripts(dir) {
 
   var results = [];
   files.forEach(function(filename) {
-    // ignore index.js and files prefixed with underscore
-    if ((filename === 'index.js') || (filename[0] === '_')) {
+    // ignore files prefixed with underscore
```
Comment (PradnyaBaviskar) [inline lib/compiler.js]: When looking for boot scripts in boot directories, the previous code, had ignored ```index.js```. Now that we resolve relative paths of boot directories, before searching for boot scripts, I wonder, if we still must be ignoring ```index.js```. 

If, per the previous code, ```index.js``` is ignored here, then for a scenario like 
```app.boot(app, { appRootDir: __dirname, bootScripts: ["./custom-boot/script" ] })```, where ```script``` is a directory, ```findScripts()``` won't be able to explore ```__dirname+"/custom-boot/script/index.js"```

So, I have modified code here to stop  ignoring ```index.js```.

@bajtos - let me know otherwise.

Comment (bajtos) [inline lib/compiler.js]: I guess it's ok to include `custom-boot/script/index.js`. However, we should keep ignoring first-level index.js files like `custom-boot/index.js`, `boot/index.js`, etc. in order to preserve backward compatibility.

**Comment by bajtos** [comment]
Please add tests for the following two scenarios a fix the implementation if needed.

```js
// 1)
appdir.writeFileSync('node_modules/custom-boot/init.js', '...');
instructions.bootScripts = 'custom-boot/init.js';
// -> file is resolved to node_modules/custom-boot/init.js

// 2)
appdir.writeFileSync('./custom-boot/init.js', '...');
appdir.writeFileSync('node_modules/custom-boot/init.js', '...');
instructions.bootScripts = 'custom-boot/init.js';
// -> file is resolved to {appRootDir}/custom-boot/init.js
// This is not great, but needed to preserve backwards compatibility
```

File: lib/compiler.js (hunk 7 above)
Comment (bajtos) [inline lib/compiler.js]: I think you may have a flaw in your reasoning. In the example you described, `findScripts` is called on `boot` only, the entry `./custom-boot/scripts` from `bootScripts` is passed directly to `resolveRelativePaths` as far as I can tell.

**Comment by PradnyaBaviskar** [comment]
Done - Added 2 test cases.

@bajtos  - Can you review?

File: test/compiler.test.js [hunk 8]
```diff
@@ -439,6 +439,86 @@ describe('compiler', function() {
...
+    it('explores index.js in `bootDirs`', function() {
+      appdir.createConfigFilesSync();
+      var initJs = appdir.writeFileSync('custom-boot/index.js',
+        'module.exports = function(app) { app.fnCalled = true; };');
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        bootDirs:['./custom-boot']
+      });
+      expect(instructions.files.boot).to.eql([initJs]);
+    });
```
Comment (bajtos) [inline test/compiler.test.js]: The index files in the directories listed in `bootDirs` must be skipped, that's the current behaviour. Please rework the test and the implementation to preserve backward compatibility.

File: test/compiler.test.js [hunk 9]
```diff
@@ -439,6 +439,86 @@ describe('compiler', function() {
...
+    it('resolves module relative path in `bootScripts`', function() {
+      appdir.createConfigFilesSync();
+      var initJs = appdir.writeFileSync('node_modules/custom-boot/init.js',
+        'module.exports = function(app) { app.fnCalled = true; };');
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        bootScripts: ['custom-boot/init.js']
+      });
+      expect(instructions.files.boot).to.eql([initJs]);
+    });
```
Comment (bajtos) [inline test/compiler.test.js]: This tests seems to be the same as L486-L495, am I missing somthing?

File: test/compiler.test.js [hunk 10]
```diff
@@ -439,6 +439,86 @@ describe('compiler', function() {
...
+      var appJs = appdir.writeFileSync('./custom-boot/init.js',
+        'module.exports = function(app) { app.fnCalled = true; };');
+      var moduleJs = appdir.writeFileSync('node_modules/custom-boot/init.js',
+        'module.exports = function(app) { app.fnCalled = true; };');
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        bootScripts: ['custom-boot/init.js']
+      });
+      expect(instructions.files.boot).to.not.eql([moduleJs]);
+      expect(instructions.files.boot).to.eql([appJs]);
```
Comment (bajtos) [inline test/compiler.test.js]: Please keep only the second assertion (expectation) line. When the array has only a single item "appJs", then it cannot be an array with a single item "moduleJs", so it's not needed to test that.

**Comment by PradnyaBaviskar** [comment]
@bajtos,
Made corrections, per above comments!

Can you review now?

File: lib/compiler.js [hunk 11]
```diff
@@ -149,8 +149,8 @@ function findScripts(dir) {
...
+    if (filename === 'index.js' || filename[0] === '_') {
```
Comment (bajtos) [inline lib/compiler.js]: Please fix the comment on L152 too.

File: test/compiler.test.js [hunk 12]
```diff
@@ -439,6 +439,76 @@ describe('compiler', function() {
...
+      var initJs = appdir.writeFileSync('custom-boot/init.js',
+        'module.exports = function(app) { app.fnCalled = true; };');
+      var instructions = boot.compile({
+        appRootDir: appdir.PATH,
+        bootScripts:['./custom-boot/init']
+      });
+      expect(instructions.files.boot).to.eql([initJs]);
+    });
+
+    it('fails to explore index.js in `bootDirs`', function() {
```
Comment (bajtos) [inline test/compiler.test.js]: Perhaps "it('ignores index.js in `bootDirs`')" would be a better name?

**Comment by bajtos** [comment]
Two more minor comments, the patch is otherwise good to be merged.

I found one more edge case that is not covered by the tests now:

```js
     var initJs = appdir.writeFileSync('node_modules/custom-boot/init.js',
       'module.exports = function(app) { app.fnCalled = true; };');
     var instructions = boot.compile({
       appRootDir: appdir.PATH,
       bootScripts: ['custom-boot/init']  // missing extension!
     });
     expect(instructions.files.boot).to.eql([initJs]);
```

I am ok to leave that out for a new pull request though.

**Comment by bajtos** [comment]
Could you please simplify the new unit-tests you are adding and create the boot scripts with empty content? They are never loaded/executed, so it does not matter what code is inside.

```diff
- appdir.writeFileSync('custom-boot/init.js',
-   'module.exports = function(app) { app.fnCalled = true; };');
+ appdir.writeFileSync('custom-boot/init.js', '');
```

**Comment by PradnyaBaviskar** [comment]
@bajtos,
Modified the test cases, so that boot scripts are loaded with empty contents. 

However I haven't yet included the edge test case, explained by you above.

Can you please review?

**Comment by bajtos** [comment]
Landed, thank you!
